torch_agents/cogment_verse_torch_agents/selfplay_td3/selfplay_td3.py

```diff
@@ -0,0 +1,180 @@
+        done = torch.FloatTensor(data["player_done"]).to(device)
+
+        return state, grid, action, reward, next_state, next_grid, done
+
+    def train(self, alice):
+        self.total_it += 1
+        training_batch = self._replay_buffer.sample()
+        state, grid, action, reward, next_state, next_grid, done = self.prepare_data(training_batch)
+
+        with torch.no_grad():
+            # Select action according to policy and add clipped noise
+            noise = (
+                torch.randn_like(action) * self._params["policy_noise"]
+            ).clamp(-self._params["noise_clip"], self._params["noise_clip"])
+
+            next_action = (
+                self._actor_target_network(next_state, next_grid) + noise
+            ).clamp(-self._params["max_action"], self._params["max_action"])
+
+            # Compute the target Q value
+            target_Q1, target_Q2 = self._critic_target_network(next_state, next_action, next_grid)
+            target_Q = torch.min(target_Q1, target_Q2)
+            # print(target_Q.shape, not_done.shape)
+            target_Q = reward.reshape(target_Q.shape) + (1.0 - done) * self._params["discount_factor"] * target_Q
+
+        # Get current Q estimates
+        current_Q1, current_Q2 = self._critic_network(state, action, grid)
+
+        # Compute critic loss
+        critic_loss = F.mse_loss(current_Q1, target_Q) + F.mse_loss(current_Q2, target_Q)
+
+        # Optimize the critic
+        self._critic_optimizer.zero_grad()
+        critic_loss.backward()
+        self._critic_optimizer.step()
+
+        # Delayed policy updates
+        if self.total_it % self._params["policy_freq"] == 0:
+
+            # Compute actor loss
+            actions = self._actor_network(state, grid)
+            actor_loss = -self._critic_network.Q1(state, actions, grid).mean()
+            if self._params["name"] == "bob":
+                # print(state.shape)
+                alice_actions = alice._actor_network(state[:, :7], grid[:, :2, :, :])
+                actor_loss = actor_loss + 0.05 * F.mse_loss(actions, alice_actions).mean()
+
+            # Optimize the actor
+            self._actor_optimizer.zero_grad()
+            actor_loss.backward()
+            self._actor_optimizer.step()
+
+            # Update the frozen target models
+            for param, target_param in zip(self._critic_network.parameters(), self._critic_target_network.parameters()):
+                target_param.data.copy_(self._params["tau"] * param.data + (1 - self._params["tau"]) * target_param.data)
+
+            for param, target_param in zip(self._actor_network.parameters(), self._actor_target_network.parameters()):
+                target_param.data.copy_(self._params["tau"] * param.data + (1 - self._params["tau"]) * target_param.data)
+
+            return actor_loss.item(), critic_loss.item() / self._params["batch_size"]
+        return 0, critic_loss.item() / self._params["batch_size"]
+
+    def learn(self, alice=None):
```

Comment:
I'm a bit confused on why this function exists. Isn't `train` handling everything? and shouldn't we just do 1 training step every time? Why do we need a for loop for params["num_training_steps"]

Reply:
Since with more "num_training_steps" the convergence improved, I needed a way to encapsulate "train" within the "num_training_steps" loop without making the "learn" function too large.